chaoscenter/graphql/server/pkg/environment/handler/handler_test.go

```diff
@@ -0,0 +1,170 @@
+	if m.GetEnvironmentsFunc != nil {
+		return m.GetEnvironmentsFunc(ctx, query)
+	}
+	return nil, nil
+}
+
+func (m *MockEnvironmentOperator) UpdateEnvironment(ctx context.Context, query bson.D, update bson.D) error {
+	if m.UpdateEnvironmentFunc != nil {
+		return m.UpdateEnvironmentFunc(ctx, query, update)
+	}
+	return nil
+}
+
+func (m *MockEnvironmentOperator) GetAggregateEnvironments(pipeline mongo.Pipeline) (*mongo.Cursor, error) {
+	if m.GetAggregateEnvironmentsFunc != nil {
+		return m.GetAggregateEnvironmentsFunc(pipeline)
+	}
+	return nil, nil
+}
+
+func (m *MockEnvironmentOperator) GetEnvironmentWithProjectID(projectID string) ([]*environments.Environment, error) {
+	if m.GetEnvironmentWithProjectIDFunc != nil {
+		return m.GetEnvironmentWithProjectIDFunc(projectID)
+	}
+	return nil, nil
+}
+
+func (m *MockEnvironmentOperator) GetEnvironmentDetails(ctx context.Context, environmentID string, projectID string) (environments.Environment, error) {
+	if m.GetEnvironmentDetailFunc != nil {
+		return m.GetEnvironmentDetailFunc(ctx, environmentID, projectID)
+	}
+	return environments.Environment{}, nil
+}
+
+func TestCreateEnvironment(t *testing.T) {
+	testCases := []struct {
+		name           string
+		projectID      string
+		input          *model.CreateEnvironmentRequest
+		token          string
+		mockInsertFunc func(ctx context.Context, env environments.Environment) error
+		expectedEnv    *model.Environment
+		expectedErr    error
+	}{
+		{
+			name:      "Failed environment creation due to invalid token",
+			projectID: "testProject",
+			input: &model.CreateEnvironmentRequest{
+				EnvironmentID: "testEnvID",
+				Name:          "testName",
+			},
+			token: "invalidToken",
+			mockInsertFunc: func(ctx context.Context, env environments.Environment) error {
+				return nil
+			},
+			expectedEnv: nil,
+			expectedErr: errors.New("invalid Token"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.WithValue(context.Background(), authorization.AuthKey, tc.token)
+			mockOperator := &MockEnvironmentOperator{
+				InsertEnvironmentFunc: tc.mockInsertFunc,
+			}
+			service := handler.NewEnvironmentService(mockOperator)
+			//service := handler.NewEnvironmentService(operator)
+
+			env, err := service.CreateEnvironment(ctx, tc.projectID, tc.input)
+			if (err != nil && tc.expectedErr == nil) ||
+				(err == nil && tc.expectedErr != nil) ||
+				(err != nil && tc.expectedErr != nil && err.Error() != tc.expectedErr.Error()) {
+				t.Fatalf("Expected error %v, but got %v", tc.expectedErr, err)
+			}
+
+			if tc.expectedEnv != nil && (env.EnvironmentID != tc.expectedEnv.EnvironmentID ||
+				env.Name != tc.expectedEnv.Name) {
+				t.Fatalf("Expected environment %+v, but got %+v", tc.expectedEnv, env)
+			}
+		})
+	}
+}
+
+func TestDeleteEnvironment(t *testing.T) {
+	testCases := []struct {
```

Comment:
Can you also add positive case too?

Reply:
@namkyu1999, the issue with positive test cases is I will need a valid token to test it. I could still test it, but I will have to mock the handler himself, but that will not be considered while running the test coverage. What do you think?

```go
type Authorizer interface {
    GetUsername(token string) (string, error)
}
type EnvironmentService struct {
	EnvironmentOperator dbOperationsEnvironment.EnvironmentOperatorInterface
	Auth                Authorizer
}

func NewEnvironmentService(EnvironmentOperator dbOperationsEnvironment.EnvironmentOperatorInterface, Auth Authorizer) EnvironmentHandler {
	return &EnvironmentService{
		EnvironmentOperator: EnvironmentOperator,
		Auth: Auth,
	}
}

func (e *EnvironmentService) CreateEnvironment(ctx context.Context, projectID string, input *model.CreateEnvironmentRequest) (*model.Environment, error) {

	currentTime := time.Now()
	if input.Tags == nil || len(input.Tags) == 0 {
		input.Tags = []string{}
	}

	tkn := ctx.Value(authorization.AuthKey).(string)
	username, err := e.Auth.GetUsername(tkn)
``` 
I tried to add an Authorizer interface but to mock it in the test but this approach doesn't on invalid test cases

Reply:
Hey @Freedisch , Here's my suggestion. What do you think?

```go
import (
        // ...
	"github.com/golang-jwt/jwt"
	"github.com/google/uuid"
	"github.com/litmuschaos/litmus/chaoscenter/graphql/server/pkg/authorization"
	"github.com/litmuschaos/litmus/chaoscenter/graphql/server/pkg/environment/handler"
	"github.com/litmuschaos/litmus/chaoscenter/graphql/server/utils"
)
const JwtSecret = "testSecret"

// getSignedJWT generates the JWT Token for the user object
func getSignedJWT(name string) (string, error) {
	token := jwt.New(jwt.SigningMethodHS512)
	claims := token.Claims.(jwt.MapClaims)
	claims["uid"] = uuid.NewString()
	claims["role"] = uuid.NewString()
	claims["username"] = name
	claims["exp"] = time.Now().Add(time.Minute).Unix()

	tokenString, err := token.SignedString([]byte(JwtSecret))
	if err != nil {
		return "", err
	}

	return tokenString, nil
}

func TestDeleteEnvironment(t *testing.T) {
	utils.Config.JwtSecret = JwtSecret
	testCases := []struct {
		name                   string
		projectID              string
		environmentID          string
		mockGetEnvironmentFunc func(query bson.D) (environments.Environment, error)
		mockUpdateFunc         func(ctx context.Context, query bson.D, update bson.D) error
		expectedResult         string
		expectedErr            error
		given                  func() string
	}{
		{
			name:          "success",
			projectID:     "testProject",
			environmentID: "testEnvID",
			mockGetEnvironmentFunc: func(query bson.D) (environments.Environment, error) {
				return environments.Environment{
					EnvironmentID: "testEnvID",
				}, nil
			},
			mockUpdateFunc: func(ctx context.Context, query bson.D, update bson.D) error {
				return nil
			},
			expectedErr: errors.New("invalid Token"),
			given: func() string {
				token, err := getSignedJWT("testUser")
				if err != nil {
					return token
				}
				return "invalid Token"
			},
		},
		{
			name:          "Failed environment deletion due to invalid token",
			projectID:     "testProject",
			environmentID: "testEnvID",
			mockGetEnvironmentFunc: func(query bson.D) (environments.Environment, error) {
				return environments.Environment{
					EnvironmentID: "testEnvID",
				}, nil
			},
			mockUpdateFunc: func(ctx context.Context, query bson.D, update bson.D) error {
				return nil
			},
			expectedErr: errors.New("invalid Token"),
			given: func() string {
				return "invalid Token"
			},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			token := tc.given()
			ctx := context.WithValue(context.Background(), authorization.AuthKey, token)

			mockOperator := &MockEnvironmentOperator{
				UpdateEnvironmentFunc: tc.mockUpdateFunc,
			}
			service := handler.NewEnvironmentService(mockOperator)

			_, err := service.DeleteEnvironment(ctx, tc.projectID, tc.environmentID)
			if (err != nil && tc.expectedErr == nil) ||
				(err == nil && tc.expectedErr != nil) ||
				(err != nil && tc.expectedErr != nil && err.Error() != tc.expectedErr.Error()) {
				t.Fatalf("Expected error %v, but got %v", tc.expectedErr, err)
			}
		})
	}
}
```

Reply:
@namkyu1999 , thanks for that. This helped complete adding positive test cases.

Reply:
I was considering mocking the token 😅, but your approach is straightforward and simple 🙌.